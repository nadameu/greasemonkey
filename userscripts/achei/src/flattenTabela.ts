export function* flattenTabela(node: Node) {
  if (node instanceof HTMLTableElement)
    yield* node.querySelector('td:nth-child(2)')?.childNodes ?? [];
  yield node;
}
