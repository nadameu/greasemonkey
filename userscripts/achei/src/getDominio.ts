const dominios = {
  '1': 'trf4',
  '2': 'jfrs',
  '3': 'jfsc',
  '4': 'jfpr',
} as Record<'1' | '2' | '3' | '4', Dominio>;

declare const OpaqueSymbol: unique symbol;
declare class OpaqueClass<T> {
  private [OpaqueSymbol]: T;
}
type Opaque<T, U> = T & OpaqueClass<U>;

declare const DominioSymbol: unique symbol;
export type Dominio = Opaque<string, { [DominioSymbol]: never }>;

export function getDominio(doc: Document) {
  const value = doc.querySelector<HTMLInputElement>(
    'input[name="local"]:checked'
  )?.value;
  if (!value || !((k): k is keyof typeof dominios => k in dominios)(value)) {
    throw new Error('Não foi possível verificar o domínio.');
  }
  return dominios[value];
}
