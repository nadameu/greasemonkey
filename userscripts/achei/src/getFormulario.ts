import { throwError } from './throwError';

export const getFormulario = async (doc: Document) =>
  doc.querySelector<HTMLFormElement>('form[name="formulario"]') ??
  throwError('Não foi possível obter o formulário.');
