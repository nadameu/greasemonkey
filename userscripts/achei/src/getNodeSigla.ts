import { NodeSigla } from './NodeSigla';
import { siglaFromText } from './siglaFromText';

export function getNodeSigla(node: Node): NodeSigla | null {
  const text = node.textContent;
  if (!text) return null;
  const sigla = siglaFromText(text);
  if (!sigla) return null;
  return { node, sigla };
}
