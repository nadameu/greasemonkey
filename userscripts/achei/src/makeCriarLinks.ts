import { NodeSigla } from './NodeSigla';

export function makeCriarLinks(doc: Document, dominio: string) {
  const template = doc.createElement('template');
  template.innerHTML = ' [ <a href="" target="_blank">Abrir na Intra</a> ]';
  const content = template.content;
  const link = content.querySelector('a')!;
  return ({ node, sigla }: NodeSigla) => {
    link.href = `https://intra.trf4.jus.br/membros/${sigla}${dominio}-jus-br`;
    const fragment = doc.importNode(content, true);
    node.parentNode!.insertBefore(fragment, node.nextSibling);
  };
}
