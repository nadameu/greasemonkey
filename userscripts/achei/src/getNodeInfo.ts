import { flattenTabela } from './flattenTabela';
import { getNodeSigla } from './getNodeSigla';
import { NodeSigla } from './NodeSigla';
import { siblings } from './siblings';

export function getNodeInfo(formulario: HTMLFormElement) {
  return Array.from(siblings(formulario))
    .flatMap(flattenTabela)
    .map(getNodeSigla)
    .filter((x): x is NodeSigla => x !== null);
}
