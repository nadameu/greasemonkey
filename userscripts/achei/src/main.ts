import { makeCriarLinks } from './makeCriarLinks';
import { getNodeInfo } from './getNodeInfo';
import { getFormulario } from './getFormulario';
import { getDominio } from './getDominio';

export async function main({ doc, log }: { doc: Document; log: typeof console.log }) {
  const dominio = getDominio(doc);
  if (!dominio) throw new Error('Não foi possível verificar o domínio.');
  const formulario = getFormulario(doc);
  if (!formulario) throw new Error('Não foi possível obter o formulário.');
  const criarLinks = makeCriarLinks(doc, dominio);
  const nodeInfo = getNodeInfo(formulario);
  for (const nodeSigla of nodeInfo) {
    criarLinks(nodeSigla);
  }
  const qtd = nodeInfo.length;
  const s = qtd > 1 ? 's' : '';
  log(`${qtd} link${s} criado${s}`);
}
