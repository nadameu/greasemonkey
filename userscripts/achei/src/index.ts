import { main } from './main';

main({ doc: document, log: console.log.bind(console, '[achei]') }).catch(
  err => {
    if (err && err instanceof Error) {
      console.error(err);
    } else {
      console.error(String(err));
    }
  }
);
