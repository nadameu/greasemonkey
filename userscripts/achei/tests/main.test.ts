import { describe, expect, MockedFunction, test, vitest } from 'vitest';
import { main } from '../src/main';

function createTest(
  name: string,
  fn: () => (
    result: PromiseSettledResult<void>,
    log: MockedFunction<{ (...args: any[]): void }>
  ) => void
) {
  return test(name, async () => {
    document.body.innerHTML = /* html */ `
<input type="radio" name="local" value="1" checked />
<form name="formulario"></form>
`;
    const after = fn();
    const log = vitest.fn();
    return Promise.allSettled([main({ doc: document, log })]).then(([x]) =>
      after(x, log)
    );
  });
}

describe('main', () => {
  createTest('sem local e sem formulario', () => {
    document.getElementsByName('local')[0]?.remove();
    document.getElementsByName('formulario')[0]?.remove();
    return (resultado, log) => {
      expect(resultado.status).toBe('rejected');
    };
  });
  createTest('sem local', () => {
    document.getElementsByName('local')[0]?.remove();

    return (resultado, log) => {
      expect(resultado.status).toBe('rejected');
    };
  });

  createTest('local com valor desconhecido', () => {
    const input = document.getElementsByName('local')[0] as HTMLInputElement;
    input.value = '999';

    return (resultado, log) => {
      expect(resultado.status).toBe('rejected');
    };
  });

  createTest('sem formulÃ¡rio', () => {
    document.getElementsByName('formulario')[0]?.remove();

    return (resultado, log) => {
      expect(resultado.status).toBe('rejected');
    };
  });

  createTest('nenhum nome', () => {
    return (resultado, log) => {
      expect(resultado.status).toBe('fulfilled');
      expect(log).toHaveBeenCalledWith('0 link criado');
    };
  });

  createTest('um nome', () => {
    const form = document.getElementsByName('formulario')[0] as HTMLFormElement;
    form.insertAdjacentHTML(
      'afterend',
      /* html */ `<br>nonononon<br>nonononon<br>Sigla: ABC00<br>nonononon<br>nonononon<br>`
    );

    return (resultado, log) => {
      expect(resultado.status).toBe('fulfilled');
      expect(log).toHaveBeenCalledWith('1 link criado');
      expect(document.body).toMatchSnapshot();
    };
  });

  createTest('dois nomes', () => {
    const form = document.getElementsByName('formulario')[0] as HTMLFormElement;
    form.insertAdjacentHTML(
      'afterend',
      /* html */ `<br>
nonononon<br>
nonononon<br>
Sigla: ABC00<br>
nonononon<br>
nonononon<br>
Sigla: bcd00 (antiga: bcd)<br>
nonononon<br>
nonononon<br>
`
    );

    return (resultado, log) => {
      expect(resultado.status).toBe('fulfilled');
      expect(log).toHaveBeenCalledWith('2 links criados');
      expect(document.body).toMatchSnapshot();
    };
  });

  createTest('com tabela', () => {
    const form = document.getElementsByName('formulario')[0] as HTMLFormElement;
    form.insertAdjacentHTML(
      'afterend',
      /* html */ `<table><tr><td><br></td><td>
nonononon<br>
nonononon<br>
Sigla: ABC00<br>
nonononon<br>
nonononon<br>
Sigla: bcd00 (antiga: bcd)<br>
nonononon<br>
nonononon<br></td></tr></table>
`
    );

    return (resultado, log) => {
      expect(resultado.status).toBe('fulfilled');
      expect(log).toHaveBeenCalledWith('2 links criados');
      expect(document.body).toMatchSnapshot();
    };
  });

  createTest('com tabela vazia', () => {
    const form = document.getElementsByName('formulario')[0] as HTMLFormElement;
    form.insertAdjacentHTML(
      'afterend',
      /* html */ `<table><tr><td><br></td></tr></table>
`
    );

    return (resultado, log) => {
      expect(resultado.status).toBe('fulfilled');
      expect(log).toHaveBeenCalledWith('0 link criado');
      expect(document.body).toMatchSnapshot();
    };
  });

  describe('dominios', () => {
    createDominioTest('JFPR', '4');
    createDominioTest('JFRS', '2');
    createDominioTest('JFSC', '3');
    function createDominioTest(name: string, value: string) {
      createTest(name, () => {
        const local = document.getElementsByName(
          'local'
        )[0] as HTMLInputElement;
        local.value = value;
        const form = document.getElementsByName(
          'formulario'
        )[0] as HTMLFormElement;
        form.insertAdjacentHTML(
          'afterend',
          /* html */ `<table><tr><td><br></td><td>
nonononon<br>
nonononon<br>
Sigla: ABC00<br>
nonononon<br>
nonononon<br>
Sigla: bcd00 (antiga: bcd)<br>
nonononon<br>
nonononon<br></td></tr></table>
`
        );

        return (resultado, log) => {
          expect(resultado.status).toBe('fulfilled');
          expect(log).toHaveBeenCalledWith('2 links criados');
          expect(document.body).toMatchSnapshot();
        };
      });
    }
  });
});
