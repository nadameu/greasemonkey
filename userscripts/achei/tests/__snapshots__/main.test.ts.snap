// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`main > com tabela 1`] = `
<body>
  

  <input
    checked=""
    name="local"
    type="radio"
    value="1"
  />
  

  <form
    name="formulario"
  />
  <table>
    <tbody>
      <tr>
        <td>
          <br />
        </td>
        <td>
          
nonononon
          <br />
          
nonononon
          <br />
          
Sigla: ABC00
           [ 
          <a
            href="https://intra.trf4.jus.br/membros/abc00trf4-jus-br"
            target="_blank"
          >
            Abrir na Intra
          </a>
           ]
          <br />
          
nonononon
          <br />
          
nonononon
          <br />
          
Sigla: bcd00 (antiga: bcd)
           [ 
          <a
            href="https://intra.trf4.jus.br/membros/bcd00trf4-jus-br"
            target="_blank"
          >
            Abrir na Intra
          </a>
           ]
          <br />
          
nonononon
          <br />
          
nonononon
          <br />
        </td>
      </tr>
    </tbody>
  </table>
  

  

</body>
`;

exports[`main > com tabela vazia 1`] = `
<body>
  

  <input
    checked=""
    name="local"
    type="radio"
    value="1"
  />
  

  <form
    name="formulario"
  />
  <table>
    <tbody>
      <tr>
        <td>
          <br />
        </td>
      </tr>
    </tbody>
  </table>
  

  

</body>
`;

exports[`main > dois nomes 1`] = `
<body>
  

  <input
    checked=""
    name="local"
    type="radio"
    value="1"
  />
  

  <form
    name="formulario"
  />
  <br />
  
nonononon
  <br />
  
nonononon
  <br />
  
Sigla: ABC00
   [ 
  <a
    href="https://intra.trf4.jus.br/membros/abc00trf4-jus-br"
    target="_blank"
  >
    Abrir na Intra
  </a>
   ]
  <br />
  
nonononon
  <br />
  
nonononon
  <br />
  
Sigla: bcd00 (antiga: bcd)
   [ 
  <a
    href="https://intra.trf4.jus.br/membros/bcd00trf4-jus-br"
    target="_blank"
  >
    Abrir na Intra
  </a>
   ]
  <br />
  
nonononon
  <br />
  
nonononon
  <br />
  

  

</body>
`;

exports[`main > dominios > JFPR 1`] = `
<body>
  

  <input
    checked=""
    name="local"
    type="radio"
    value="4"
  />
  

  <form
    name="formulario"
  />
  <table>
    <tbody>
      <tr>
        <td>
          <br />
        </td>
        <td>
          
nonononon
          <br />
          
nonononon
          <br />
          
Sigla: ABC00
           [ 
          <a
            href="https://intra.trf4.jus.br/membros/abc00jfpr-jus-br"
            target="_blank"
          >
            Abrir na Intra
          </a>
           ]
          <br />
          
nonononon
          <br />
          
nonononon
          <br />
          
Sigla: bcd00 (antiga: bcd)
           [ 
          <a
            href="https://intra.trf4.jus.br/membros/bcd00jfpr-jus-br"
            target="_blank"
          >
            Abrir na Intra
          </a>
           ]
          <br />
          
nonononon
          <br />
          
nonononon
          <br />
        </td>
      </tr>
    </tbody>
  </table>
  

  

</body>
`;

exports[`main > dominios > JFRS 1`] = `
<body>
  

  <input
    checked=""
    name="local"
    type="radio"
    value="2"
  />
  

  <form
    name="formulario"
  />
  <table>
    <tbody>
      <tr>
        <td>
          <br />
        </td>
        <td>
          
nonononon
          <br />
          
nonononon
          <br />
          
Sigla: ABC00
           [ 
          <a
            href="https://intra.trf4.jus.br/membros/abc00jfrs-jus-br"
            target="_blank"
          >
            Abrir na Intra
          </a>
           ]
          <br />
          
nonononon
          <br />
          
nonononon
          <br />
          
Sigla: bcd00 (antiga: bcd)
           [ 
          <a
            href="https://intra.trf4.jus.br/membros/bcd00jfrs-jus-br"
            target="_blank"
          >
            Abrir na Intra
          </a>
           ]
          <br />
          
nonononon
          <br />
          
nonononon
          <br />
        </td>
      </tr>
    </tbody>
  </table>
  

  

</body>
`;

exports[`main > dominios > JFSC 1`] = `
<body>
  

  <input
    checked=""
    name="local"
    type="radio"
    value="3"
  />
  

  <form
    name="formulario"
  />
  <table>
    <tbody>
      <tr>
        <td>
          <br />
        </td>
        <td>
          
nonononon
          <br />
          
nonononon
          <br />
          
Sigla: ABC00
           [ 
          <a
            href="https://intra.trf4.jus.br/membros/abc00jfsc-jus-br"
            target="_blank"
          >
            Abrir na Intra
          </a>
           ]
          <br />
          
nonononon
          <br />
          
nonononon
          <br />
          
Sigla: bcd00 (antiga: bcd)
           [ 
          <a
            href="https://intra.trf4.jus.br/membros/bcd00jfsc-jus-br"
            target="_blank"
          >
            Abrir na Intra
          </a>
           ]
          <br />
          
nonononon
          <br />
          
nonononon
          <br />
        </td>
      </tr>
    </tbody>
  </table>
  

  

</body>
`;

exports[`main > um nome 1`] = `
<body>
  

  <input
    checked=""
    name="local"
    type="radio"
    value="1"
  />
  

  <form
    name="formulario"
  />
  <br />
  nonononon
  <br />
  nonononon
  <br />
  Sigla: ABC00
   [ 
  <a
    href="https://intra.trf4.jus.br/membros/abc00trf4-jus-br"
    target="_blank"
  >
    Abrir na Intra
  </a>
   ]
  <br />
  nonononon
  <br />
  nonononon
  <br />
  

</body>
`;
